daemon/policy.go

```diff
@@ -206,27 +207,47 @@ func (d *Daemon) PolicyAdd(rules policyAPI.Rules, opts *AddOptions) (uint64, err
 	}
 
 	d.policy.Mutex.Lock()
+	// removedPrefixes tracks prefixes that we replace in the rules. It is used
+	// after we release the policy repository lock.
+	var removedPrefixes []*net.IPNet
 	if opts != nil {
 		if opts.Replace {
 			for _, r := range rules {
-				tmp := d.policy.SearchRLocked(r.Labels)
-				if len(tmp) > 0 {
-					d.dnsPoller.StopPollForDNSName(tmp)
+				oldRules := d.policy.SearchRLocked(r.Labels)
+				removedPrefixes = append(removedPrefixes, policy.GetCIDRPrefixes(oldRules)...)
+				if len(oldRules) > 0 {
+					d.dnsPoller.StopPollForDNSName(oldRules)
 					d.policy.DeleteByLabelsLocked(r.Labels)
 				}
 			}
 		}
 		if len(opts.ReplaceWithLabels) > 0 {
-			tmp := d.policy.SearchRLocked(opts.ReplaceWithLabels)
-			if len(tmp) > 0 {
-				d.dnsPoller.StopPollForDNSName(tmp)
+			oldRules := d.policy.SearchRLocked(opts.ReplaceWithLabels)
+			removedPrefixes = append(removedPrefixes, policy.GetCIDRPrefixes(oldRules)...)
```

Comment:
Actually, coming back to this... Could this end up appending 'nil' to the list of removedPrefixes and end up in some weird situation later on?

Reply:
if `policy.GetCIDRPrefixes(oldRules)` returns a nil or an empty slice, then no. This is because the `...` appends each entry in the second parameter to the first (more akin to a python `.extend` vs `.append`). If the return includes nil, though, then is will append those elements as well. Is that a possibility? I can guard against it.

Reply:
I think the code is fine, returning `[{ nil }]` would be a bug in `GetCIDRPrefixes()`